src/plugins/shell/shell_impl.cpp

```diff
@@ -0,0 +1,215 @@
+#include "shell_impl.h"
+#include "system.h"
+
+namespace mavsdk {
+
+void ShellImpl::init()
+{
+    using namespace std::placeholders; // for `_1`
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_SERIAL_CONTROL,
+        std::bind(&ShellImpl::process_shell_message, this, _1),
+        this);
+}
+
+void ShellImpl::deinit()
+{
+    _parent->unregister_all_mavlink_message_handlers(this);
+}
+
+void ShellImpl::enable() {}
+
+void ShellImpl::disable() {}
+
+ShellImpl::ShellImpl(System& system) :
+    PluginImplBase(system),
+    _transfer_finished_future(_transfer_finished_promise.get_future())
+{
+    _parent->register_plugin(this);
+    _transfer_finished_promise.set_value();
+}
+
+ShellImpl::~ShellImpl()
+{
+    _parent->unregister_plugin(this);
+}
+
+bool ShellImpl::is_transfer_in_progress()
+{
+    if (!_transfer_finished_future.valid()) {
+        return true;
+    }
+    return _transfer_finished_future.wait_for(std::chrono::seconds(0)) != std::future_status::ready;
```

Comment:
Could you use `valid()` here?
https://en.cppreference.com/w/cpp/thread/future/valid

Reply:
Actually this is not what we want.
`valid` returns `true` until `get()` is called on the future. So in our case it will return `true` every time.
And in this case 
```cpp
if (!_transfer_finished_future.valid()) {
    return true;
}
```
seems like useless check. Is should fix it :)

Reply:
Oh, I see. Nevermind then.